pkg/controller/component/helper.go

```diff
@@ -144,3 +153,98 @@ func syncConfigMap(
 	}
 	return cm, cmHash, nil
 }
+
+func getContainerImage(
+	extender extenderutil.UnstructuredExtender,
+	obj *unstructured.Unstructured,
+	containerName string) string {
+	if obj == nil {
+		return ""
+	}
+	containers := extender.GetContainers(obj)
+	for _, ctr := range containers {
+		if ctr["name"] == containerName {
+			return ctr["image"].(string)
+		}
+	}
+	return ""
+}
+
+func isUpdating(
+	component v1alpha1.NebulaClusterComponentter,
+	podClient kube.Pod,
+	obj *unstructured.Unstructured) (bool, error) {
+	extender := extenderutil.New()
+	if extenderutil.IsUpdating(extender, obj) {
+		return true, nil
+	}
+
+	selector, err := label.Label(component.GenerateLabels()).Selector()
+	if err != nil {
+		return false, err
+	}
+
+	pods, err := podClient.ListPods(component.GetNamespace(), selector)
+	if err != nil {
+		return false, fmt.Errorf(
+			"failed to get pods for cluster %s/%s, selector %s, error: %s",
+			component.GetNamespace(),
+			component.GetClusterName(),
+			selector,
+			err,
+		)
+	}
+	for i := range pods {
+		pod := pods[i]
+		revisionHash, exist := pod.Labels[appsv1.ControllerRevisionHashLabelKey]
+		if !exist {
+			return false, nil
+		}
+		if revisionHash != component.GetUpdateRevision() {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func setPartition(
+	extender extenderutil.UnstructuredExtender,
+	obj *unstructured.Unstructured,
+	upgradeOrdinal int64,
+	advanced bool,
+) error {
+	return extenderutil.SetUpdatePartition(extender, obj, upgradeOrdinal, InPlaceGracePeriodSeconds, advanced)
+}
+
+func getNextUpdatePod(component v1alpha1.NebulaClusterComponentter, replicas int32, podClient kube.Pod) (int32, error) {
+	namespace := component.GetNamespace()
+	updateRevision := component.GetUpdateRevision()
+	for index := replicas - 1; index >= 0; index-- {
+		podName := component.GetPodName(index)
+		pod, err := podClient.GetPod(namespace, podName)
+		if err != nil {
+			return -1, err
+		}
+		revision, exist := pod.Labels[appsv1.ControllerRevisionHashLabelKey]
+		if !exist {
+			return -1, err
+		}
+		if revision == updateRevision {
+			continue
+		}
+
+		return index, nil
+	}
+	return -1, nil
+}
+
+func getLastConfig(extender extenderutil.UnstructuredExtender, actual, desired *unstructured.Unstructured) error {
```

Comment:
It looks like it’s not getting the config, but updating to `desired`.
I think the function name is a bit confusing, what do you think?

Reply:
Fixed